apps/omg_child_chain/lib/omg_child_chain/fees/fee_parser.ex

```diff
@@ -11,35 +11,86 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-
 defmodule OMG.ChildChain.Fees.FeeParser do
   @moduledoc """
   Transaction's fee validation functions
   """
-
   require Logger
 
+  # the format of the json is invalid (ie: it's an array)
+  @type parsing_error() ::
+          :invalid_json_format
+          # the tx type can't be parsed to an integer
+          | :invalid_tx_type
+          # the fee spec for a specific type/token is missing keys
+          | :invalid_fee_spec
+          # the fee amount is invalid (must be >= 0)
+          | :invalid_fee
+          # the subunit to unit is invalid (must be > 0)
+          | :invalid_subunit_to_unit
+          # the pegged amount is invalid (must be > 0)
+          | :invalid_pegged_amount
+          # the pegged currency is invalid (must be > 0)
+          | :invalid_pegged_currency
+          # the pegged subunit to unit is invalid (must be > 0)
+          | :invalid_pegged_subunit_to_unit
+          # the updated at date is invalid (wrong date format)
+          | :invalid_timestamp
+          # the token address is invalid (must be a valid Ethereum address)
+          | :bad_address_encoding
+          # There is a duplicated token for the same tx type
+          | :duplicate_token
+
   @doc """
   Parses and validates json encoded fee specifications file
-
   Parses provided json string to token-fee map and returns the map together with possible parsing errors
   """
-  @spec parse(binary()) :: {:ok, OMG.Fees.fee_t()} | {:error, list({:error, atom()})}
+  @spec parse(binary()) ::
+          {:ok, OMG.Fees.full_fee_t()} | {:error, list({:error, parsing_error(), any(), non_neg_integer() | nil})}
   def parse(file_content) do
     with {:ok, json} <- Jason.decode(file_content) do
-      {errors, token_fee_map, _} =
-        json
-        |> Enum.map(&parse_fee_spec/1)
-        |> Enum.reduce({[], %{}, 1}, &spec_reducer/2)
+      {errors, fee_specs} = Enum.reduce(json, {[], %{}}, &reduce_json/2)
 
       errors
       |> Enum.reverse()
-      |> (&{&1, token_fee_map}).()
+      |> (&{&1, fee_specs}).()
       |> handle_parser_output()
     end
   end
 
-  defp parse_fee_spec(%{
+  defp reduce_json({tx_type, fee_spec}, {all_errors, fee_specs}) do
+    tx_type
+    |> Integer.parse()
+    |> parse_for_type(fee_spec)
+    |> handle_type_parsing_output(tx_type, all_errors, fee_specs)
+  end
+
+  defp reduce_json(_, {all_errors, fee_specs}) do
+    {[{:error, :invalid_json_format, nil, nil}] ++ all_errors, fee_specs}
+  end
+
+  defp parse_for_type({tx_type, ""}, fee_spec) do
+    fee_spec
+    |> Enum.map(&parse_single_spec/1)
+    |> Enum.reduce({[], %{}, 1, tx_type}, &spec_reducer/2)
```

Comment:
these seems to be rather complicated parts (lots of validation and different schema possibilities). I would suggest you do the following:
- extract parse_single_spec into a separate module and have it tested there (the validation would fit nicely there)
- spec_reducer and it's `|> Enum.reduce({[], %{}, 1, tx_type}, &spec_reducer/2)` might be easier to digest in a recursive function rather than a map reduce, it allows you to nicely specify the initial elements (accumulators)  

is carrying the error list around the parsing necessary? is that used on the API level? where does the requirement for this come from?

Reply:
Extracted the parsing of a single spec in its own module with specific tests.
All the error handling was already in the code when I added the tx type support so I didn't want to mess with it and just wanted to keep it as it is. It's not used on the API level but for logging purpose.

Reply:
Love this!